File: src/menu/menues/ClientLobbyMenu.java
package menu.menues;

import java.net.InetAddress;

import core.Main;
import game.Team;
import game.avatar.Avatar;
import game.skill.Skill;
import game.item.Item;
import misc.Debug;
import network.Packet;
import network.lobby.LobbyPlayer;
import network.lobby.packets.user.*;
import network.lobby.packets.*;

public class ClientLobbyMenu extends LobbyMenu // menu of client when in lobby
{
	private LobbyPlayer localPlayer; // Verweiß auf den eigenen LobbyPlayer
	private InetAddress serverIP; // Speichert die IP adresse des Servers

	public ClientLobbyMenu(String serverIPString)
	{
		try
		{
			this.serverIP = InetAddress.getByName(serverIPString); // Speichert die Internet-Adresse vom Server; Wird von JoinServerMenu übergeben
		} catch (Exception e) { Debug.warn("ClientLobbyMenu not a valid ip address: " + serverIPString); }
	}

	// Wird ausgeführt von Main.networkDevice.receive()
	@Override public void handlePacket(Packet packet, InetAddress ip)
	{
		if (packet instanceof LobbyPlayersPacket) // Wenn es sich bei dem Packet um die Liste aller Player handelt, dann
		{
			if (getLocalPlayer() == null)
			{
				setPlayers(((LobbyPlayersPacket) packet).getPlayers()); // setzte die komplette Liste neu
				Debug.note("ClientLobbyMenu.handlePacket(): received LobbyPlayersPacket -> eigene players neu gesetzt", Debug.Tags.LOBBYMENUINFO);
				updatePlayerIcons();
			}
			else
			{
				int i = playerToID(getLocalPlayer(), getPlayers());
				setPlayers(((LobbyPlayersPacket) packet).getPlayers()); // setzte die komplette Liste neu
				localPlayer = getPlayers().get(i);
				if (localPlayer == null)
				{
					Debug.warn("ClientLobbyMenu.handlePacket(): localPlayer == null");
				}
				updatePlayerIcons();
			}
		}
		else if (packet instanceof MapPacket)
		{
			if (getPhase() == TEAM_PHASE)
			{
				// TODO apply
				unlockAll();
			}
			else
			{
				Debug.error("client> got mappacket out of teamphase");
			}
		}
		else if (packet instanceof UserPacketWithID) // Wenn es sich bei dem Packet um UserPacketWithID handelt, dann
		{
			UserPacket userPacket = ((UserPacketWithID) packet).getUserPacket(); // entpacken des UserPackets aus dem UserpacketWithID

			if (userPacket instanceof LockUserPacket) // wenn du ein lock-packet erhalten hast
			{
				if (isServerPlayer(getPlayer(packet)))
				{
					nextPhase();
				}
				else
				{
					getPlayer(packet).applyUserPacket(userPacket);
				}
			}
			else if (userPacket instanceof DisconnectUserPacket)
			{
				if (getPlayers() == null)
				{
					Debug.warn("ClientLobbyMenu.handlePacket(): getPlayers() == null");
				}
				if (getPlayer(packet) == null)
				{
					Debug.warn("ClientLobbyMenu.handlePacket(): getPlayer(packet) == null");
				}
				getPlayers().remove(getPlayer(packet));
				unlockAll();
				updatePlayerIcons();
			}
			else // falls nicht.
			{
				switch (getPhase())
				{
					case TEAM_PHASE: // Falls man in der Team Phase ist
						if (userPacket instanceof TeamUserPacket) // Wenn Spieler das Team wechselt
						{
							if (getPlayer(packet) == null)
							{
								Debug.warn("ClientLobbyMenu.handlePacket(): getPlayer(packet) == null (2)");
							}
							getPlayer(packet).applyUserPacket(userPacket); // setzte das Team des jeweilgen Spielers
							updatePlayerIcons();
							unlockAll();
						}
						else if (userPacket instanceof LoginUserPacket) // Wenn sich ein Spieler einloggt
						{
							LobbyPlayer player = new LobbyPlayer((LoginUserPacket) userPacket); // Neuen LobbyPlayer erstellen
							if (getPlayers().size() < 1)
							{
								Debug.warn("ClientLobbyMenu.handlePacket(): getPlayers().size() == " + getPlayers().size());
							}
							getPlayers().add(player); // Zur Liste der LobbyPlayer hinzufügen
							if (localPlayer == null) // Da das erste LoginUserPacket, das man bekommt, das eigene ist
							{
								localPlayer = player; // Diesen LobbyPlayer als eigenen abspeichern
							}
							unlockAll();
							updatePlayerIcons();
						}
						else
						{
							Debug.error("Client can't accept packet (" + packet + ") in team phase"); // Alle anderen Packete werden in dieser Phase nicht angenommen
						}
						break;
					case AVATAR_PHASE: // Wenn man sich in der AvatarPhase befindet
						if (userPacket instanceof AvatarUserPacket) // Falls ein Spieler seinen Avatar wechselt
						{
							if(getPlayer(packet) == null)
							{
								Debug.warn("ClientLobbyMenu.handlePacket(): getPlayer(packet) == null (3)");
							}
							getPlayer(packet).applyUserPacket(userPacket); // Des jeweiligen Spielers Avatar wird in den LobbyPlayern geändert
						}
						else
						{
							Debug.error("Client can't accept packet (" + packet + ") in avatar phase"); // Alle anderen Packete werden in dieser Phase nicht angenommen
						}
						break;
					case SKILL_PHASE: // Wenn man in der SkillPhase ist
						if (userPacket instanceof SkillUserPacket) // Wenn ein Spieler ein Skill wechselt
						{
							getPlayer(packet).applyUserPacket(userPacket); // In der LobbyPlayer Liste übernehmen
						}
						else
						{
							Debug.error("Client can't accept packet (" + packet + ") in skill phase"); // Alle anderen Packete werden in dieser Phase nicht angenommen
						}
						break;
					case ITEM_PHASE:
						if (userPacket instanceof ItemUserPacket) // Wenn ein Spieler ein Item wechselt
						{
							getPlayer(packet).applyUserPacket(userPacket); // In der LobbyPlayer Liste übernehmen
						}
						else
						{
							Debug.error("Client can't accept packet (" + packet + ") in item phase"); // Alle anderen Packete werden in dieser Phase nicht angenommen
						}
						break;
					default:
						Debug.error("ClientLobbyMenu.handlePacket(...): wrong phase (" + getPhase() + ")"); // Da ist was ganz komisch gelaufen; Ungültige Phase
						break;
				}
			}
		}
		else
		{
			Debug.error("Client received wrong packet (" + packet + ")"); // packets, die nicht vom Typ UserPacketWithID sind werden nicht angenommen
		}
	}

	// Wird ausgeführt, wenn ChoosePlayerPropertyMenu mit okButton geschlossen wird
	@Override public void sendPlayerPropertyUpdate()
	{
		if (getLocalPlayer() == null)
		{
			Debug.warn("ClientLobbyMenu.sendPlayerPropertyUpdate(): getLocalPlayer == null");
		}
		switch (getPhase())
		{
			case AVATAR_PHASE:
				sendToServer(getLocalPlayer().getAvatarPacket());
				break;
			case SKILL_PHASE:
				sendToServer(getLocalPlayer().getSkillPacket());
				break;
			case ITEM_PHASE:
				sendToServer(getLocalPlayer().getItemPacket());
				break;
			default:
				Debug.error("ClientLobbyMenu.sendPlayerPropertyUpdate(): wrong phase: " + getPhase());
		}
	}

	// Wird aufgerufen, wenn man auf die Map clickt
	@Override public void mapPressed()
	{
		// TODO: Map anzeigen
	}

	// Wird aufgerufen, wenn man den Lock-Button drückt
	@Override public void lockPressed()
	{
		if (isPlayerPropertiesChoosen())
		{
			if (getLocalPlayer() == null)
			{
				Debug.warn("ClientLobbyMenu.lockPressed(): getLocalPlayer() == null");
			}
			sendToServer(new LockUserPacket(!getLocalPlayer().isLocked())); // sendet an den Server, dass man den lock-button betätigt hat
		}
		else
		{
			Debug.note("ClientLobbyMenu.lockPressed: Kann noch nicht locken, da noch nicht alle PlayerProperties gewählt wurden", Debug.Tags.LOBBYMENUINFO);
		}
	}

	@Override public void disconnectPressed()
	{
		sendToServer(new DisconnectUserPacket());
		Main.getMenues().remove(Main.getMenues().getLast());
		Debug.note("ClientLobbyMenu.disconnectPressed: LobbyMenu beendet", Debug.Tags.LOBBYMENUINFO);
	}

	// Wird aufgerufen, wenn man sein Team wechselt
	@Override public void teamPressed(Team team)
	{
		if (getLocalPlayer() == null)
		{
			Debug.warn("ClientLobbyMenu.teamPressed(): getLocalPlayer() == null");
		}
		if (!getLocalPlayer().isLocked() && !getLocalPlayer().getTeam().equals(team))
		{
			sendToServer(new TeamUserPacket(team)); // An die anderen Spieler senden
		}
	}


	@Override public void avatarPressed(Avatar avatar)
	{
		// if new avatar TODO
		// sendToServer(new AvatarUserPacket(avatar)); // TODO
	}

	@Override public void skillPressed(Skill[] skills)
	{
		// if new skills TODO
		// sendToServer(new SkillUserPacket(skills));
	}

	@Override public void itemPressed(Item[] items)
	{
		// if new items TODO
		// sendToServer(new ItemUserPacket(items));
	}

	@Override public LobbyPlayer getLocalPlayer()
	{
		if (localPlayer == null)
		{
			Debug.warn("ClientLobbyMenu.getLocalPlayer(): getLocalPlayer return null");
		}
		return localPlayer;
	}

	// returnt den Spieler, der den das übergebene Packet anspricht
	private LobbyPlayer getPlayer(Packet packet)	
	{
		if (packet == null)
		{
			Debug.error("ClientLobbyMenu.getPlayer(): packet == null");
		}
		UserPacketWithID userPacketWithID = (UserPacketWithID) packet;
		return getPlayers().get(userPacketWithID.getID());
	}

	// Sendet das Packet zum Server
	private void sendToServer(Packet packet)
	{
		send(packet, serverIP);
	}

	private boolean isServerPlayer(LobbyPlayer player)
	{
		return player == getServerPlayer();
	}

	private LobbyPlayer getServerPlayer()
	{
		if (getPlayers() == null)
			Debug.warn("ClientLobbyMenu.getServerPlayer(): getPlayers() == null");
		if (getPlayers().get(0) == null)
		{
			Debug.warn("ClientLobbyMenu.getServerPlayer(): returns null");
		}
		return getPlayers().get(0);
	}
}

File: cpp/src/menu/menues/ClientLobbyMenu.cpp
#include "ClientLobbyMenu.hpp"

#include <misc/Global.hpp>
#include <menu/MenuList.hpp>
#include <menu/components/Button.hpp>
#include <core/Main.hpp>
#include <game/ClientGameInterface.hpp>
#include <network/packets/LoginPacket.hpp>
#include <network/packets/SkillPacket.hpp>
#include <network/packets/AvatarPacket.hpp>
#include <network/packets/ItemPacket.hpp>
#include <network/packets/PlayerPropertyPacket.hpp>
#include <network/packets/MapPacket.hpp>
#include <network/packets/LobbyPlayersPacket.hpp>
#include <network/packets/TeamPacket.hpp>
#include <network/packets/DisconnectPacket.hpp>
#include <network/packets/LockPacket.hpp>
#include <network/packets/PacketWithID.hpp>
#include <network/packets/GameStartPacket.hpp>
#include <tilemap/LobbyTileMap.hpp>
#include <player/LobbyPlayer.hpp>
#include <team/Team.hpp>

ClientLobbyMenu::ClientLobbyMenu(const std::string& ip)
	: LobbyMenu("Lock"), localPlayerID(-1)
{
	serverIP = ip;
	LoginPacket* p = new LoginPacket(Main::getName(), Main::getRank());
	sendToServer(p);
	deleteAndNullptr(p);
}

ClientLobbyMenu::~ClientLobbyMenu()
{}

LobbyPlayer* ClientLobbyMenu::getLocalPlayer() const
{
	return getPlayer(localPlayerID);
}

void ClientLobbyMenu::handlePacket(Packet* packet, const sf::IpAddress& ip)
{
	if (packet == nullptr) Debug::warn("ClientLobbyMenu::handlePacket(): packet == nullptr");
	if (ip == serverIP)
	{
		handlePacketByID(packet, 0);
	}
	else
	{
		Debug::warn("got packet from non-server player");
		deleteAndNullptr(packet);
	}
}

void ClientLobbyMenu::handlePacketByID(Packet* packet, int id)
{
	if (packet->getCompressID() == CompressIDs::PACKET_WITH_ID)
	{
		PacketWithID* packetWithID = packet->unwrap<PacketWithID>();
		handlePacketByID(packetWithID->getPacket(), packetWithID->getID());
	}
	else if (packet->getCompressID() == CompressIDs::LOCK_PACKET)
	{
		handleLockPacket(packet->unwrap<LockPacket>(), id);
	}
	else if (packet->getCompressID() == CompressIDs::DISCONNECT_PACKET)
	{
		handleDisconnectPacket(packet->unwrap<DisconnectPacket>(), id);
	}
	else if (packet->getCompressID() == CompressIDs::TEAM_PACKET && getPhase() == TEAM_PHASE)
	{
		handleTeamPacket(packet->unwrap<TeamPacket>(), id);
	}
	else if (packet->getCompressID() == CompressIDs::LOGIN_PACKET && getPhase() == TEAM_PHASE)
	{
		handleLoginPacket(packet->unwrap<LoginPacket>());
	}
	else if (packet->getCompressID() == CompressIDs::AVATAR_PACKET && getPhase() == AVATAR_PHASE)
	{
		handleAvatarPacket(packet->unwrap<AvatarPacket>(), id);
	}
	else if (packet->getCompressID() == CompressIDs::SKILL_PACKET && getPhase() == SKILL_PHASE)
	{
		handleSkillPacket(packet->unwrap<SkillPacket>(), id);
	}
	else if (packet->getCompressID() == CompressIDs::ITEM_PACKET && getPhase() == ITEM_PHASE)
	{
		handleItemPacket(packet->unwrap<ItemPacket>(), id);
	}
	else if (packet->getCompressID() == CompressIDs::MAP_PACKET && getPhase() == TEAM_PHASE)
	{
		handleMapPacket(packet->unwrap<MapPacket>());
	}
	else if (packet->getCompressID() == CompressIDs::LOBBY_PLAYERS_PACKET)
	{
		handleLobbyPlayersPacket(packet->unwrap<LobbyPlayersPacket>());
	}
	else if (packet->getCompressID() == CompressIDs::GAME_START_PACKET)
	{
		handleGameStartPacket(packet->unwrap<GameStartPacket>());
	}
	else
	{
		Debug::warn("ClientLobbyMenu::handlePacketByID(): awkward packet(" + Converter::intToString((int) packet->getCompressID()) + ") in awkward phase(" + Converter::intToString(getPhase()) + ")");
	}
	deleteAndNullptr(packet);
}

void ClientLobbyMenu::lockPressed()
{
	// Error-Handling
	if (getLocalPlayer() == nullptr) Debug::warn("ClientLobbyMenu::lockPressed(): getLocalPlayer() == nullptr");
	if (getLocalPlayer()->getLockPacket() == nullptr) Debug::warn("ClientLobbyMenu::lockPressed(): getLocalPlayer()->getLockPacket() == nullptr");

	// send changed lock status to server (not applied here)
	LockPacket* packet = new LockPacket(!getLocalPlayer()->getLockPacket()->isLocked());
	sendToServer(packet);
	deleteAndNullptr(packet);
}

void ClientLobbyMenu::disconnectPressed()
{
	DisconnectPacket* packet = new DisconnectPacket();
	sendToServer(packet);
	deleteAndNullptr(packet);
	Main::getMenuList()->back();
}

void ClientLobbyMenu::teamPressed(Team* team)
{
	TeamPacket* packet = new TeamPacket(team->getID());
	sendToServer(packet);
	deleteAndNullptr(packet);
}

void ClientLobbyMenu::sendToServer(Packet* p)
{
	send(p, serverIP);
}

void ClientLobbyMenu::handleLoginPacket(LoginPacket* packet)
{
	LobbyPlayer* player = new LobbyPlayer(packet);
	addPlayer(player);
	if (localPlayerID == -1)
	{
		localPlayerID = getPlayers().size()-1;
	}

	unlockAll();
	updatePlayerIcons();
}

void ClientLobbyMenu::handleLobbyPlayersPacket(LobbyPlayersPacket* packet)
{
	clearPlayers();
	std::vector<LobbyPlayer*> players(packet->getPlayers());
	for (unsigned int i = 0; i < players.size(); i++)
	{
		addPlayer(players[i]);
	}

	if (global::AUTO_LOBBY)
	{
		switch (getPhase())
		{
			case AVATAR_PHASE:
			{
				playerPropertySelected(new AvatarPacket(global::AUTO_AVATAR));
				break;
			}
			case SKILL_PHASE:
			{
				playerPropertySelected(new SkillPacket(global::AUTO_SKILLS));
				break;
			}
			case ITEM_PHASE:
			{
				playerPropertySelected(new ItemPacket(global::AUTO_ITEMS));
				break;
			}
		}
	}

	updatePlayerIcons();
}

void ClientLobbyMenu::handleLockPacket(LockPacket* packet, int id)
{
	if (id == 0)
	{
		nextPhase();
	}
	else
	{
		getPlayer(id)->applyLockPacket(packet);
	}
}

void ClientLobbyMenu::handleDisconnectPacket(DisconnectPacket* packet, int id)
{
	removePlayer(id);

	unlockAll();
	updatePlayerIcons();
}

void ClientLobbyMenu::handleTeamPacket(TeamPacket* packet, int id)
{
	getPlayer(id)->applyTeamPacket(packet);

	unlockAll();
	updatePlayerIcons();
}

void ClientLobbyMenu::handleAvatarPacket(AvatarPacket* packet, int id)
{
	LobbyPlayer* player = getPlayer(id);
	if (player == nullptr) Debug::warn("ClientLobbyMenu::handleAvatarPacket(): getPlayer(" + Converter::intToString(id) + ") == nullptr");
	player->applyAvatarPacket(packet);
}

void ClientLobbyMenu::handleSkillPacket(SkillPacket* packet, int id)
{
	LobbyPlayer* player = getPlayer(id);
	if (player == nullptr) Debug::warn("ClientLobbyMenu::handleSkillPacket(): getPlayer(" + Converter::intToString(id) + ") == nullptr");
	player->applySkillPacket(packet);
}

void ClientLobbyMenu::handleItemPacket(ItemPacket* packet, int id)
{

	LobbyPlayer* player = getPlayer(id);
	if (player == nullptr) Debug::warn("ClientLobbyMenu::handleItemPacket(): getPlayer(" + Converter::intToString(id) + ") == nullptr");
	player->applyItemPacket(packet);
}

void ClientLobbyMenu::handleMapPacket(MapPacket* packet)
{
	unlockAll();
	updateMap(packet->getTileIDs());
}

void ClientLobbyMenu::handleGameStartPacket(GameStartPacket* packet)
{
	Main::getMenuList()->addMenu(new ClientGameInterface(getLobbyTileMap(), getPlayers(), localPlayerID, serverIP, packet->getUnixBeginTime()));
	resetLobby();
}

void ClientLobbyMenu::updateLockButton() const
{
	switch (getPhase())
	{
		case TEAM_PHASE:
			lockButton->setEnabled(getLobbyTileMap()->isValid());
		break;
		case AVATAR_PHASE:
			lockButton->setEnabled(getLocalPlayer()->getAvatarPacket()->isValid());
		break;
		case SKILL_PHASE:
			lockButton->setEnabled(getLocalPlayer()->getSkillPacket()->isValid());
		break;
		case ITEM_PHASE:
			lockButton->setEnabled(getLocalPlayer()->getItemPacket()->isValid());
		break;
	}
}

void ClientLobbyMenu::playerPropertySelected(PlayerPropertyPacket* packet)
{
	switch (packet->getCompressID())
	{
		case CompressIDs::AVATAR_PACKET:
			getLocalPlayer()->applyAvatarPacket(packet->unwrap<AvatarPacket>());
			break;
		case CompressIDs::SKILL_PACKET:
			getLocalPlayer()->applySkillPacket(packet->unwrap<SkillPacket>());
			break;
		case CompressIDs::ITEM_PACKET:
			getLocalPlayer()->applyItemPacket(packet->unwrap<ItemPacket>());
			break;
		default:
			Debug::warn("ClientLobbyMenu::playerPropertySelected(): awkward packet");
	}
	sendToServer(packet);
}
